File: carin/parser/iso.py
"""Minimal ISO 9660 reader + CARINdb virtual volume mapper.

Reads the navigation image directly (no mount required) and exposes the
CARINdb block address space that spans DB_0 + DB_1.

Reference: docs/CARINDB_BLUEPRINT.md sections 1 and 2.
"""

from __future__ import annotations

import struct
import zlib
from dataclasses import dataclass
from typing import Dict, Iterator, Optional

ISO_SECTOR = 2048
CARIN_SECTOR = 512
CARIN_WINDOW = 0x400000  # CARINdb sectors covered by one DB_n file (2 GiB)

BLOCK_HDR = ">IHBB"
BLOCK_HDR_SIZE = 8
SECTION_DESC = ">HH"


@dataclass(frozen=True)
class IsoFile:
    path: str
    lba: int
    size: int

    @property
    def offset(self) -> int:
        return self.lba * ISO_SECTOR


class IsoImage:
    """Just enough ISO 9660 to locate file extents by path."""

    def __init__(self, path: str):
        self._f = open(path, "rb")
        self.files: Dict[str, IsoFile] = {}
        self._read_pvd()

    def _read_pvd(self) -> None:
        self._f.seek(16 * ISO_SECTOR)
        pvd = self._f.read(ISO_SECTOR)
        if pvd[0] != 1 or pvd[1:6] != b"CD001":
            raise ValueError("not an ISO 9660 primary volume descriptor")
        self.volume_id = pvd[40:72].decode("latin-1").strip()
        self.publisher = pvd[318:446].decode("latin-1").strip()
        root = pvd[156:190]
        root_lba = struct.unpack_from("<I", root, 2)[0]
        root_size = struct.unpack_from("<I", root, 10)[0]
        self._walk(root_lba, root_size, "")

    def _walk(self, lba: int, size: int, prefix: str) -> None:
        self._f.seek(lba * ISO_SECTOR)
        data = self._f.read(size)
        pos = 0
        while pos < len(data):
            rec_len = data[pos]
            if rec_len == 0:
                # padding to the next logical sector
                pos = (pos // ISO_SECTOR + 1) * ISO_SECTOR
                if pos >= len(data):
                    break
                continue
            rec = data[pos:pos + rec_len]
            ext_lba = struct.unpack_from("<I", rec, 2)[0]
            ext_size = struct.unpack_from("<I", rec, 10)[0]
            flags = rec[25]
            name_len = rec[32]
            name = rec[33:33 + name_len].decode("latin-1")
            pos += rec_len
            if name_len == 1 and name in ("\x00", "\x01"):
                continue  # '.' and '..'
            name = name.split(";")[0]
            full = f"{prefix}/{name}"
            if flags & 0x02:
                self._walk(ext_lba, ext_size, full)
            else:
                self.files[full] = IsoFile(full, ext_lba, ext_size)

    def read(self, iso_file: IsoFile, offset: int, length: int) -> bytes:
        if offset >= iso_file.size:
            return b""
        length = min(length, iso_file.size - offset)
        self._f.seek(iso_file.offset + offset)
        return self._f.read(length)

    def close(self) -> None:
        self._f.close()


@dataclass
class CarinBlock:
    sector: int          # absolute virtual CARINdb sector
    length: int          # on-disk length, in 512-byte sectors
    type: int
    comp: int            # 0 = raw, 1 = unknown codec, 2 = zlib
    usize: int           # decompressed size, in 512-byte sectors
    raw: bytes           # on-disk bytes, header included
    data: Optional[bytes]  # decompressed bytes, header included (None if comp == 1)

    @property
    def payload(self) -> bytes:
        if self.data is None:
            raise ValueError(f"block {self.sector} uses COMPRESSION_FLAG=1 (unknown codec)")
        return self.data

    def sections(self, n: int):
        return [
            struct.unpack_from(SECTION_DESC, self.payload, BLOCK_HDR_SIZE + 4 * i)
            for i in range(n)
        ]


class CarinVolume:
    """CARINdb block address space spanning DB_0 .. DB_n."""

    def __init__(self, image: IsoImage, db_paths=("/DB/DB_0", "/DB/DB_1")):
        self.image = image
        self.parts = [image.files[p] for p in db_paths]
        self.sectors = [f.size // CARIN_SECTOR for f in self.parts]

    def read_sectors(self, sector: int, count: int) -> bytes:
        idx, local = divmod(sector, CARIN_WINDOW)
        if idx >= len(self.parts):
            raise ValueError(f"sector {sector} outside volume")
        return self.image.read(self.parts[idx], local * CARIN_SECTOR, count * CARIN_SECTOR)

    def block(self, sector: int) -> CarinBlock:
        head = self.read_sectors(sector, 1)
        bid, btype, cf, us = struct.unpack_from(BLOCK_HDR, head)
        if (bid >> 8) != sector:
            raise ValueError(f"sector {sector}: BLOCK_ID says {bid >> 8}")
        length = bid & 0xFF
        raw = head if length == 1 else self.read_sectors(sector, length)
        if cf == 0:
            data = raw
        elif cf == 2:
            data = raw[:BLOCK_HDR_SIZE] + zlib.decompress(raw[BLOCK_HDR_SIZE:])
        else:
            data = None
        return CarinBlock(sector, length, btype, cf, us, raw, data)

    def walk(self, part: Optional[int] = None) -> Iterator[CarinBlock]:
        """Follow the block chain. Yields header-only blocks (raw not sliced)."""
        parts = range(len(self.parts)) if part is None else [part]
        for idx in parts:
            base = idx * CARIN_WINDOW
            nsec = self.sectors[idx]
            sector = 0
            buf, buf_base = b"", -1
            CHUNK = 1 << 22
            while sector < nsec:
                off = sector * CARIN_SECTOR
                if buf_base < 0 or off < buf_base or off + BLOCK_HDR_SIZE > buf_base + len(buf):
                    buf = self.image.read(self.parts[idx], off, CHUNK)
                    buf_base = off
                    if len(buf) < BLOCK_HDR_SIZE:
                        break
                bid, btype, cf, us = struct.unpack_from(BLOCK_HDR, buf, off - buf_base)
                length = bid & 0xFF
                if (bid >> 8) != base + sector or length == 0:
                    sector += 1
                    continue
                yield CarinBlock(base + sector, length, btype, cf, us, b"", None)
                sector += length


# --------------------------------------------------------------------------
# Geographic frame (docs/CARINDB_BLUEPRINT.md section 7)
# A full 360 deg turn is 2_000_000_000 CARIN units; the origin sits on the
# equator at 30 deg West.  Calibrated against 38 city labels, rms 2.0 km.
# --------------------------------------------------------------------------

CARIN_UNITS_PER_TURN = 2_000_000_000
K = CARIN_UNITS_PER_TURN / 360.0          # 5_555_555.5555... units per degree
LON_ORIGIN = -30.0
LAT_ORIGIN = 0.0

# plausible coverage window, used to recognise a bbox inside service data
_X_RANGE = (-60_000_000, 1_000_000_000)
_Y_RANGE = (0, 450_000_000)
QUADTREE_UNIT = 98_304          # smallest observed tile side, in CARIN units


def to_wgs84(x: int, y: int):
    return (x / K + LON_ORIGIN, y / K + LAT_ORIGIN)


def to_carin(lon: float, lat: float):
    return (round((lon - LON_ORIGIN) * K), round((lat - LAT_ORIGIN) * K))


def find_bbox(payload: bytes):
    """Locate the 4 x int32 bounding box that follows the section descriptor.

    The descriptor entry count varies per block, so the bbox is found by
    scanning 4-aligned positions up to the first section offset.  A genuine box
    sits on the CARIN quadtree: both sides are multiples of 98304 units and the
    aspect ratio is 1:1, 1:2 or 2:1.
    Returns (offset, (x0, y0, x1, y1)) or None.
    """
    if len(payload) < 0x30:
        return None
    first = struct.unpack_from(">H", payload, BLOCK_HDR_SIZE)[0]
    limit = min(first if first > 16 else len(payload), len(payload) - 16, 512)
    for off in range(12, limit, 4):
        x0, y0, x1, y1 = struct.unpack_from(">4i", payload, off)
        if not (x0 < x1 and y0 < y1):
            continue
        if not (_X_RANGE[0] <= x0 and x1 <= _X_RANGE[1]):
            continue
        if not (_Y_RANGE[0] <= y0 and y1 <= _Y_RANGE[1]):
            continue
        w, h = x1 - x0, y1 - y0
        if w % QUADTREE_UNIT or h % QUADTREE_UNIT:
            continue
        if w not in (h, 2 * h) and h != 2 * w:
            continue
        if w > 200_000_000:
            continue
        return off, (x0, y0, x1, y1)
    return None

File: carindb-rs/src/iso.rs
use std::collections::HashMap;
use std::fs::File;
use std::io;
use std::path::Path;
use memmap2::{Mmap, MmapOptions};

pub const ISO_SECTOR: usize = 2048;
pub const CARIN_SECTOR: usize = 512;
pub const CARIN_WINDOW: usize = 0x400000;
pub const BLOCK_HDR_SIZE: usize = 8;

pub struct IsoImage {
    mmap: Mmap,
    pub files: HashMap<String, IsoFile>,
}

#[derive(Clone, Debug)]
pub struct IsoFile {
    pub path: String,
    pub lba: u32,
    pub size: u32,
}

impl IsoImage {
    pub fn new<P: AsRef<Path>>(path: P) -> io::Result<Self> {
        let file = File::open(path)?;
        let mmap = unsafe { MmapOptions::new().map(&file)? };
        
        let mut iso = IsoImage {
            mmap,
            files: HashMap::new(),
        };
        iso.read_pvd()?;
        Ok(iso)
    }

    fn read_pvd(&mut self) -> io::Result<()> {
        let pvd_offset = 16 * ISO_SECTOR;
        if self.mmap[pvd_offset] != 1 || &self.mmap[pvd_offset + 1..pvd_offset + 6] != b"CD001" {
            return Err(io::Error::new(io::ErrorKind::InvalidData, "Not a valid ISO 9660 PVD"));
        }

        let root_dir_record = &self.mmap[pvd_offset + 156..pvd_offset + 190];
        let root_lba = u32::from_le_bytes(root_dir_record[2..6].try_into().unwrap());
        let root_size = u32::from_le_bytes(root_dir_record[10..14].try_into().unwrap());

        let root_files = self.walk(root_lba, root_size, "")?;
        self.files = root_files;
        Ok(())
    }

    fn walk(&self, lba: u32, size: u32, prefix: &str) -> io::Result<HashMap<String, IsoFile>> {
        let mut files = HashMap::new();
        let start = (lba as usize) * ISO_SECTOR;
        let end = start + (size as usize);
        let mut pos = start;

        while pos < end {
            let rec_len = self.mmap[pos] as usize;
            if rec_len == 0 {
                pos = (pos / ISO_SECTOR + 1) * ISO_SECTOR;
                continue;
            }
            if pos + 33 >= end { break; }

            let name_len = self.mmap[pos + 32] as usize;
            let name_bytes = &self.mmap[pos + 33..pos + 33 + name_len];
            let mut name = String::from_utf8_lossy(name_bytes).to_string();
            
            if name_len == 1 && (name == "\x00" || name == "\x01") {
                pos += rec_len;
                continue;
            }
            
            if let Some(semi) = name.find(';') {
                name.truncate(semi);
            }
            
            let full = if prefix.is_empty() {
                format!("/{}", name)
            } else {
                format!("{}/{}", prefix, name)
            };

            let flags = self.mmap[pos + 25];
            let ext_lba = u32::from_le_bytes(self.mmap[pos + 2..pos + 6].try_into().unwrap());
            let ext_size = u32::from_le_bytes(self.mmap[pos + 10..pos + 14].try_into().unwrap());

            if (flags & 0x02) != 0 {
                let sub_files = self.walk(ext_lba, ext_size, &full)?;
                files.extend(sub_files);
            } else {
                files.insert(full.clone(), IsoFile {
                    path: full,
                    lba: ext_lba,
                    size: ext_size,
                });
            }
            
            pos += rec_len;
        }

        Ok(files)
    }

    pub fn read_slice(&self, file: &IsoFile, offset: usize, length: usize) -> &[u8] {
        let file_start = (file.lba as usize) * ISO_SECTOR;
        let start = file_start + offset;
        let end = std::cmp::min(start + length, file_start + file.size as usize);
        &self.mmap[start..end]
    }
}

pub struct CarinVolume {
    pub image: IsoImage,
    parts: Vec<IsoFile>,
    sectors: Vec<usize>,
    layout: Option<HashMap<u16, u16>>,
    pub db_rel: u16,
}

pub struct CarinBlock<'a> {
    pub sector: u32,
    pub length: u8,
    pub btype: u16,
    pub comp: u8,
    pub usize_: u8,
    pub raw: &'a [u8],
}

impl CarinVolume {
    pub fn new(image: IsoImage) -> io::Result<Self> {
        let db_0 = image.files.get("/DB/DB_0").cloned().ok_or_else(|| io::Error::new(io::ErrorKind::NotFound, "DB_0 not found"))?;
        let db_1 = image.files.get("/DB/DB_1").cloned().ok_or_else(|| io::Error::new(io::ErrorKind::NotFound, "DB_1 not found"))?;
        let parts = vec![db_0.clone(), db_1.clone()];
        let sectors = vec![
            (db_0.size as usize) / CARIN_SECTOR,
            (db_1.size as usize) / CARIN_SECTOR,
        ];
        
        let mut vol = CarinVolume {
            image,
            parts,
            sectors,
            layout: None,
            db_rel: 0,
        };
        
        vol.read_layout();
        
        Ok(vol)
    }

    fn read_layout(&mut self) {
        let mut layout = HashMap::new();
        let db_rel;
        {
            let sb = self.read_sectors(0, 2);
            let off = u16::from_be_bytes(sb[0x28..0x2A].try_into().unwrap()) as usize;
            let count = u16::from_be_bytes(sb[0x2A..0x2C].try_into().unwrap()) as usize;
            
            for i in 0..count {
                let p = off + i * 4;
                let key = u16::from_be_bytes(sb[p..p+2].try_into().unwrap());
                let val = u16::from_be_bytes(sb[p+2..p+4].try_into().unwrap());
                layout.insert(key, val);
            }
            db_rel = u16::from_be_bytes(sb[0x1A..0x1C].try_into().unwrap());
        }
        self.layout = Some(layout);
        self.db_rel = db_rel;
    }

    pub fn layout(&self) -> &HashMap<u16, u16> {
        self.layout.as_ref().unwrap()
    }

    pub fn read_sectors(&self, sector: usize, count: usize) -> &[u8] {
        let idx = sector / CARIN_WINDOW;
        let local = sector % CARIN_WINDOW;
        self.image.read_slice(&self.parts[idx], local * CARIN_SECTOR, count * CARIN_SECTOR)
    }

    pub fn block(&self, sector: usize) -> Option<CarinBlock> {
        let head = self.read_sectors(sector, 1);
        if head.len() < BLOCK_HDR_SIZE { return None; }
        
        let bid = u32::from_be_bytes(head[0..4].try_into().unwrap());
        let btype = u16::from_be_bytes(head[4..6].try_into().unwrap());
        let comp = head[6];
        let usize_ = head[7];
        
        let length = (bid & 0xFF) as usize;
        if length == 0 { return None; }
        
        let raw = self.read_sectors(sector, length);
        
        Some(CarinBlock {
            sector: sector as u32,
            length: length as u8,
            btype,
            comp,
            usize_,
            raw,
        })
    }
}

pub struct BlockIter<'a> {
    vol: &'a CarinVolume,
    current_part: usize,
    current_sector: usize,
}

impl<'a> CarinVolume {
    pub fn iter_blocks(&'a self) -> BlockIter<'a> {
        BlockIter {
            vol: self,
            current_part: 0,
            current_sector: 0,
        }
    }
}

impl<'a> Iterator for BlockIter<'a> {
    type Item = CarinBlock<'a>;

    fn next(&mut self) -> Option<Self::Item> {
        while self.current_part < self.vol.parts.len() {
            let base = self.current_part * CARIN_WINDOW;
            let nsec = self.vol.sectors[self.current_part];
            
            while self.current_sector < nsec {
                let absolute_sector = base + self.current_sector;
                let head = self.vol.read_sectors(absolute_sector, 1);
                if head.len() < 8 {
                    self.current_sector += 1;
                    continue;
                }
                
                let bid = u32::from_be_bytes(head[0..4].try_into().unwrap());
                let btype = u16::from_be_bytes(head[4..6].try_into().unwrap());
                let comp = head[6];
                let usize_ = head[7];
                
                let length = (bid & 0xFF) as usize;
                
                if (bid >> 8) != absolute_sector as u32 || length == 0 {
                    self.current_sector += 1;
                    continue;
                }
                
                let raw = self.vol.read_sectors(absolute_sector, length);
                
                self.current_sector += length;
                
                return Some(CarinBlock {
                    sector: absolute_sector as u32,
                    length: length as u8,
                    btype,
                    comp,
                    usize_,
                    raw,
                });
            }
            self.current_part += 1;
            self.current_sector = 0;
        }
        None
    }
}
